lib/utils/polling_client.js

```diff
@@ -0,0 +1,234 @@
+  }
+
+  if (!interval || interval < 0) {
+    interval = self.defaultPollingIntervalMs;
+  }
+
+  if (interval < self.minPollingIntervalMs) {
+    interval = self.minPollingIntervalMs;
+  }
+
+  return interval;
+};
+
+pollingClient.prototype.startIntervalPolling = function startIntervalPolling() {
+  var self = this;
+
+  if (self.timeoutId) {
+    return;
+  }
+
+  self.pollStateOnce(self.token, function onPoll(err) {
+    if (err) {
+      self.handleErrors(err);
+      return;
+    }
+
+    self.intervalPoll();
+  });
+};
+
+pollingClient.prototype.intervalPoll = function intervalPoll() {
+  var self = this;
+
+  clearTimeout(self.timeoutId);
+
+  var interval = self.calcPollingInterval();
+
+  // We are not using interval so we consider that the interval / tokens
+  // etc could have changed
+  self.timeoutId = setTimeout(function onPoll() {
+    self.pollStateOnce(self.token);
+
+    if (!self.timeoutId) {
+      return;
+    }
+
+    self.intervalPoll(self.token);
+  }, interval);
+};
+
+pollingClient.prototype.emitEvents = function emitEvents(err, oldState, newState) {
+  var self = this;
+
+  if (err) {
+    self.hub.emit('error', err);
+    return;
+  }
+
+  if (oldState && oldState.state === 'pending' &&
+  !oldState.enrollment && newState.enrollment) {
+    self.hub.emit('enrollment:confirmed', {
+      device_account: newState.enrollment,
+      tx_id: newState.id
+    });
+  }
+
+  if (newState.state === 'accepted') {
+    self.hub.emit('login:complete', {
+      tx_id: newState.id,
+      signature: newState.token
+    });
+
+    self.closeConnection();
+  }
+
+  if (newState.state === 'rejected') {
+    self.hub.emit('login:rejected', { tx_id: newState.id });
+  }
+};
+
+/**
+ *
+ */
+pollingClient.prototype.closeConnection = function closeConnection() {
+  clearTimeout(this.timeoutId);
+  delete this.timeoutId;
+};
+
+pollingClient.prototype.connect = function connect(token, callback) {
+  this.token = token;
+  this.startIntervalPolling(token);
+  callback();
+};
+
+pollingClient.prototype.addListener = function addListener(kind, event, handler) {
+  var wrapped = wrapEventHandler(handler);
+
+  listenersMap.set(handler, wrapped);
+  this.hub[kind](event, wrapped);
+};
+
+pollingClient.prototype.on = function on(event, handler) {
+  this.addListener('on', event, handler);
+};
+
+pollingClient.prototype.once = function once(event, handler) {
+  this.addListener('once', event, handler);
+};
+
+pollingClient.prototype.removeListener = function removeListener(event, handler) {
+  var wrapped = listenersMap.get(handler);
+
+  this.hub.removeListener(event, wrapped);
+  listenersMap.del(handler);
+};
+
+pollingClient.prototype.removeAllListeners = function removeAllListeners(eventName) {
+  this.hub.removeAllListeners(eventName);
+};
+
+function wrapEventHandler(handler) {
+  return function wrapper(payload) {
+    handler(object.toCamelKeys(payload));
+  };
+}
```

Comment:
Wouldn't it be simpler (and also lighter for the cpu) to handle payload transformation in `emitEvents`, and not have to wrap and keep a store of wrapped handlers?  ...or even use camel cased keys in emitEvents and just have to camelize error from remote?